### PR TITLE
Add configurable limit for max # connections per server (TVD-912)

### PR DESCRIPTION
**Pull request by tillberg**
This modifies the memcached client to have a configurable limit for the maximum number of connections per remote server address; it does this by using https://golang.org/pkg/sync/#Cond with its `Wait` and `Signal` functions to build a map of counting semaphores. It's not efficient for large number of memcached servers (since there is a single shared Cond across all remote addresses), but it should perform fine for a smallish number of servers (i.e. less than a dozen or so).

As implemented, the max idle connections and max active connections settings are package-wide and are both per-address limits, not global limits across all addresses.

### CODE REVIEW

File: memcache/memcache.go
Comment (ggerrietts) [inline memcache/memcache.go:198]: Deleting these lines is unrelated to the change under review, yes?

File: memcache/memcache.go
Comment (ggerrietts) [inline memcache/memcache.go]: Pretty sure you meant "signal" not "single".

Comment (ggerrietts) [reply memcache/memcache.go]: Might want to split this comment along the abstraction lines that you've split the code.

**Review by ggerrietts — APPROVED** [review]
I think you should probably maintain that comment but I'll leave it up to you.

File: memcache/memcache.go
Comment (tillberg) [reply memcache/memcache.go:198]: I moved this initialization up to the `NewFromSelector` function at https://github.com/tracelytics/gomemcache/pull/1/files#diff-789b976323c7e2b437c8c37b5674e7b8R132.

I guess that the base version of this code was written such that you could get away with instantiating your own `Client` without using the `New*` functions, but I added two more properties that need to be initialized for `Client`, and lazy initialization of all those things starts to get kind of verbose. Kind of a bikeshed change, but _shrug_... I'm not planning to push for this patch to be merged upstream, as that seems kind of intentionally barebones (also, while there are other forks, none have seemed to have the need to implement a max connection limit).

**Comment by tillberg** [comment]
I realized yesterday that there was significant bug in the previous implementation. Signal() only wakes up a single caller blocked on Wait(); but that the caller that gets woken up might be waiting for an slot to become available for a _different_ address, and so it would call Wait() again without Signal()ing another goroutine to wake up and try. I'm not sure if this could deadlock entirely, but it would at least dramatically under-utilize available connection slots when there is more than one memcached server address. It could have been patched over by using Broadcast() instead, but that would scale very poorly to large numbers of addresses.

So I changed the implementation to use a channel per address. Channels because that makes it easier to add a timeout waiting for a slot to open up, and Per Address because that makes this scale better to large numbers of addresses and also fixes the bug described in the previous paragraph.

**Review by ggerrietts — APPROVED** [review]
This strikes me as a little easier to read than the previous incarnation.

I had a kneejerk reaction to the mutation of the channel objects outside the mutex, but then I realized that channels are designed as concurrency primitives, and I've just been thinking of them too narrowly.

Also, I think that while you add a decent amount of code, your code is not really all that invasive. It will probably merge pretty easily.

**Comment by cce** [comment]
Did you look at https://github.com/fatih/pool? Looks like a similar approach

**Pull request merged**
